apis/metal3.io/v1alpha1/firmwareschema_types.go

```diff
@@ -54,85 +56,114 @@ type SettingSchema struct {
 	Unique *bool `json:"unique,omitempty"`
 }
 
-// FirmwareSchemaSpec defines the desired state of FirmwareSchema
-type FirmwareSchemaSpec struct {
-
-	// The hardware vendor associated with this schema
-	// +optional
-	HardwareVendor string `json:"hardwareVendor,omitempty"`
-
-	// The hardware model associated with this schema
-	// +optional
-	HardwareModel string `json:"hardwareModel,omitempty"`
-
-	// Map of firmware name to schema
-	Schema map[string]SettingSchema `json:"schema" required:"true"`
+type SchemaSettingError struct {
+	name    string
+	message string
 }
 
-//+kubebuilder:object:root=true
-
-// FirmwareSchema is the Schema for the firmwareschemas API
-type FirmwareSchema struct {
-	metav1.TypeMeta   `json:",inline"`
-	metav1.ObjectMeta `json:"metadata,omitempty"`
-
-	Spec FirmwareSchemaSpec `json:"spec,omitempty"`
+func (e SchemaSettingError) Error() string {
+	return fmt.Sprintf("Setting %s is invalid, %s", e.name, e.message)
 }
 
-// Check whether the setting's name and value is valid using the schema
-func (host *FirmwareSchema) CheckSettingIsValid(name string, value intstr.IntOrString, schemas map[string]SettingSchema) bool {
-
-	schema, ok := schemas[name]
-	if !ok {
-		// The setting must exist in the status
-		return false
-	}
+func (schema *SettingSchema) IsValid(name string, value intstr.IntOrString) error {
 
 	if schema.ReadOnly != nil && *schema.ReadOnly == true {
-		return false
+		return SchemaSettingError{name: name, message: "it is ReadOnly"}
 	}
 
 	// Check if valid based on type
 	switch schema.AttributeType {
 	case "Enumeration":
 		for _, av := range schema.AllowableValues {
 			if value.String() == av {
-				return true
+				return nil
 			}
 		}
-		return false
+		return SchemaSettingError{name: name, message: fmt.Sprintf("unknown enumeration value - %s", value.String())}
 
 	case "Integer":
-		if schema.LowerBound == nil || schema.UpperBound == nil {
+		if schema.LowerBound == nil && schema.UpperBound == nil {
 			// return true if no settings to check validity
```

Comment:
this comment probably needs to be changed

Reply:
Fixed